flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeStore.java

```diff
@@ -0,0 +1,173 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.lang.Thread.holdsLock;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.apache.flink.util.ExceptionUtils.findThrowable;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A {@link StateChangeStore} that waits for some configured amount of time before passing the
+ * accumulated state changes to the actual store. The writes are executed asynchronously.
+ */
+@ThreadSafe
+class BatchingStateChangeStore implements StateChangeStore {
+    private static final Logger LOG = LoggerFactory.getLogger(BatchingStateChangeStore.class);
+
+    private final StateChangeStore delegate;
+    private final ScheduledExecutorService scheduler;
+    private final long scheduleDelayMs;
+    private final long sizeThresholdBytes;
+
+    @GuardedBy("scheduled")
+    private final Queue<StoreTask> scheduled;
+
+    @GuardedBy("scheduled")
+    private long scheduledSizeInBytes;
+
+    @GuardedBy("scheduled")
+    private Future<?> scheduledFuture = CompletableFuture.completedFuture(null);
+
+    private AtomicReference<Throwable> error = new AtomicReference<>(null);
+
+    BatchingStateChangeStore(
+            long persistDelayMs, long sizeThresholdBytes, StateChangeStore delegate) {
+        this(
+                persistDelayMs,
+                sizeThresholdBytes,
+                delegate,
+                SchedulerFactory.create(1, "ChangelogRetryScheduler", LOG));
+    }
+
+    BatchingStateChangeStore(
+            long persistDelayMs,
+            long sizeThresholdBytes,
+            StateChangeStore delegate,
+            ScheduledExecutorService scheduler) {
+        this.scheduleDelayMs = persistDelayMs;
+        this.scheduled = new LinkedList<>();
+        this.scheduler = scheduler;
+        this.sizeThresholdBytes = sizeThresholdBytes;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void save(StoreTask storeTask) {
+        Collection<StateChangeSet> changeSets = storeTask.changeSets;
+        if (error.get() != null) {
+            LOG.debug("don't persist {} changesets, already failed", changeSets.size());
+            storeTask.fail(error.get());
+            return;
+        }
+        LOG.debug("persist {} changeSets", changeSets.size());
+        try {
+            synchronized (scheduled) {
+                scheduled.add(storeTask);
+                scheduledSizeInBytes += storeTask.getSize();
+                scheduleUploadIfNeeded();
+            }
+        } catch (Exception e) {
+            storeTask.fail(e);
+            throw e;
+        }
+    }
+
+    private void scheduleUploadIfNeeded() {
+        checkState(holdsLock(scheduled));
+        if (scheduledFuture.isDone() || isOverSizeThresholdAndCancellationSucceded()) {
+            scheduleUpload();
```

Comment:
I think there is a race condition here which could lead to upload not to be scheduled:
1. Future is runnig in `drainAndSave()`, enters `scheduleUploadIfNeeded()` and sees no need to reschedule (no data added)
2. It exits the `synchronized` section but not the task yet (so future isn't finished yet)
3. Another thread adds data and sees future.isDone == false, so it doesn't schedule neither

WDYT?

(the related issue below is about `scheduledFuture.cancel`, this one is about `scheduledFuture.isDone()`; I guess fix will address both of them)